### PR TITLE
Create rolebinding for the PR author.

### LINKED ISSUE #65: Let the GitHub PR Author have view (or edit?) access into the namespace created by their tests
This helps people iterate on stuff if they login to the api.ci cluster

/cc @bbguimaraes @smarterclayton

### PR DESCRIPTION
**Pull request by droslean**
Create rolebinding for the PR author to give him access in the created temporary namespace.
closes #65

### CODE REVIEW

**Comment by gabemontero** [comment]
cool deal

File: cmd/ci-operator/main.go [hunk 1]
```diff
@@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
 		break
 	}
 
+	if o.givePrAuthorAccessToNamespace {
+		// Generate rolebinding for the PR Author.
+		if len(o.jobSpec.Refs.Pulls) == 1 {
```
Comment (stevekuznetsov) [inline cmd/ci-operator/main.go]:
```go
for _, pull := range o.jobSpec.Refs.Pulls {
```

File: cmd/ci-operator/main.go [hunk 2]
```diff
@@ -228,6 +230,7 @@ func bindOptions(flag *flag.FlagSet) *options {
 
 	// experimental flags
 	flag.StringVar(&opt.gitRef, "git-ref", "", "Populate the job spec from this local Git reference. If JOB_SPEC is set, the refs field will be overwritten.")
+	flag.BoolVar(&opt.givePrAuthorAccessToNamespace, "give-pr-author-access-to-namespace", true, "Give view access to the temporarily created namespace to the PR author.")
```
Comment (stevekuznetsov) [inline cmd/ci-operator/main.go]: default to `false`

File: cmd/ci-operator/main.go [hunk 3]
```diff
@@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
...
+		if len(o.jobSpec.Refs.Pulls) == 1 {
+			rbacClient, err := rbacclientset.NewForConfig(o.clusterConfig)
+			if err != nil {
+				return fmt.Errorf("could not get RBAC client for cluster config: %v", err)
+			}
+
+			log.Printf("Creating rolebinding for user %s in namespace %s", o.jobSpec.Refs.Pulls[0].Author, o.namespace)
+			if _, err := rbacClient.RoleBindings(o.namespace).Create(&rbacapi.RoleBinding{
+				ObjectMeta: meta.ObjectMeta{
+					Name:      "namespace-viewer",
```
Comment (stevekuznetsov) [inline cmd/ci-operator/main.go]: I thought we wanted `admin`? @gabemontero what did you think you would need?

**Comment by gabemontero** [comment]
On Wed, Aug 8, 2018 at 10:58 AM Steve Kuznetsov <notifications@github.com>
wrote:

> *@stevekuznetsov* commented on this pull request.
> ------------------------------
>
> In cmd/ci-operator/main.go
> <https://github.com/openshift/ci-operator/pull/80#discussion_r208615200>:
>
> > @@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
>  		break
>  	}
>
> +	if o.givePrAuthorAccessToNamespace {
> +		// Generate rolebinding for the PR Author.
> +		if len(o.jobSpec.Refs.Pulls) == 1 {
>
> for _, pull := range o.jobSpec.Refs.Pulls {
>
> ------------------------------
>
> In cmd/ci-operator/main.go
> <https://github.com/openshift/ci-operator/pull/80#discussion_r208615280>:
>
> > @@ -228,6 +230,7 @@ func bindOptions(flag *flag.FlagSet) *options {
>
>  	// experimental flags
>  	flag.StringVar(&opt.gitRef, "git-ref", "", "Populate the job spec from this local Git reference. If JOB_SPEC is set, the refs field will be overwritten.")
> +	flag.BoolVar(&opt.givePrAuthorAccessToNamespace, "give-pr-author-access-to-namespace", true, "Give view access to the temporarily created namespace to the PR author.")
>
> default to false
> ------------------------------
>
> In cmd/ci-operator/main.go
> <https://github.com/openshift/ci-operator/pull/80#discussion_r208615497>:
>
> > @@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
>  		break
>  	}
>
> +	if o.givePrAuthorAccessToNamespace {
> +		// Generate rolebinding for the PR Author.
> +		if len(o.jobSpec.Refs.Pulls) == 1 {
> +			rbacClient, err := rbacclientset.NewForConfig(o.clusterConfig)
> +			if err != nil {
> +				return fmt.Errorf("could not get RBAC client for cluster config: %v", err)
> +			}
> +
> +			log.Printf("Creating rolebinding for user %s in namespace %s", o.jobSpec.Refs.Pulls[0].Author, o.namespace)
> +			if _, err := rbacClient.RoleBindings(o.namespace).Create(&rbacapi.RoleBinding{
> +				ObjectMeta: meta.ObjectMeta{
> +					Name:      "namespace-viewer",
>
> I thought we wanted admin? @gabemontero <https://github.com/gabemontero>
> what did you think you would need?
>
yeah at least edit would be useful
I could see retriggering a build or tweaking an image stream as possible
diagnostics of use

Maybe giving someone else access to my PRs project to help debug would be a
reason to do admin

File: cmd/ci-operator/main.go (hunk 1 above)
Comment (droslean) [reply cmd/ci-operator/main.go]: In which cases there are more than 1 PRs ? Only in batch jobs right. I don't think that we need to include this case as well.

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: I don't see a great reason not to allow access to all authors in a batch. If the job is still being tested, we might want to investigate a batch run.

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: This is a nit, if you feel strongly on it we can merge as-is

File: cmd/ci-operator/main.go [hunk 4]
```diff
@@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
 		break
 	}
 
+	if o.givePrAuthorAccessToNamespace {
+		// Generate rolebinding for the PR Author.
+		if len(o.jobSpec.Refs.Pulls) == 1 {
+			rbacClient, err := rbacclientset.NewForConfig(o.clusterConfig)
```
Comment (bbguimaraes) [inline cmd/ci-operator/main.go]: More of a question than a review: would it be worth it to create the RBAC
client only once when both `--give-pr-author-access-to-namespace` and
`--delete-when-idle` are set?

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: Good follow up, let's do this. Only call the `Create()` in the loop

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: @droslean

File: cmd/ci-operator/main.go [hunk 5]
```diff
@@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
 		break
 	}
 
+	if o.givePrAuthorAccessToNamespace {
+		// Generate rolebinding for all the PR Authors.
+		for _, pull := range o.jobSpec.Refs.Pulls {
+			rbacClient, err := rbacclientset.NewForConfig(o.clusterConfig)
```
Comment (smarterclayton) [inline cmd/ci-operator/main.go]: Don’t create one client per author

File: cmd/ci-operator/main.go [hunk 6]
```diff
@@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
...
+			if err != nil {
+				return fmt.Errorf("could not get RBAC client for cluster config: %v", err)
+			}
+
+			log.Printf("Creating rolebinding for user %s in namespace %s", pull.Author, o.namespace)
+			if _, err := rbacClient.RoleBindings(o.namespace).Create(&rbacapi.RoleBinding{
+				ObjectMeta: meta.ObjectMeta{
+					Name:      "temp-namespace-admin",
```
Comment (smarterclayton) [inline cmd/ci-operator/main.go]: There’s nothing temporary about this, give it a clearer name like “ci-op-author-access”

File: cmd/ci-operator/main.go [hunk 7]
```diff
@@ -534,6 +537,31 @@ func (o *options) initializeNamespace() error {
...
+					Namespace: o.namespace,
+				},
+				Subjects: []rbacapi.Subject{{Kind: "User", Name: pull.Author}},
+				RoleRef: rbacapi.RoleRef{
+					Kind: "ClusterRole",
```
Comment (smarterclayton) [inline cmd/ci-operator/main.go]: No no no!

This gives any PR author access to our cluster provisioning secrets!

This needs to be view access only.

Comment (stevekuznetsov) [reply cmd/ci-operator/main.go]: Oh jeez this should be `Role` only ... bad review on my part

**Pull request merged**